marconi-core-json-rpc/src/Marconi/Core/JsonRpc.hs

```diff
@@ -55,14 +148,30 @@ queryIndexerVarHttpHandler indexerVar q = do
 queryIndexerHttpHandler
   :: ( Show (Core.Result query)
      , Ord (Core.Point event)
-     , Core.Queryable (ExceptT (Core.QueryError query) Handler) event query indexer
-     , Core.IsSync (ExceptT (Core.QueryError query) Handler) event indexer
+     , Core.Queryable (ExceptT (Core.QueryError query) IO) event query indexer
+     , Core.IsSync IO event indexer
      )
   => indexer event
   -> query
   -> Handler (Either (JsonRpcErr String) (Core.Result query))
-queryIndexerHttpHandler indexer q = do
-  countE <- runExceptT $ Core.queryLatest q indexer
-  case countE of
-    Left err -> pure $ Left $ mkJsonRpcInternalErr (Just $ show err)
-    Right count -> pure $ Right count
+queryIndexerHttpHandler indexer q =
+  liftIO $ first queryErrToRpcErr <$> Core.queryLatestEither q indexer
+
+-- | Convert a query error to a JSON-RPC protocol error.
+queryErrToRpcErr
+  :: (Show (Core.Result query))
+  => Core.QueryError query
+  -> JsonRpcErr String
+queryErrToRpcErr = \case
+  Core.NotStoredAnymore ->
+    mkJsonRpcInvalidParamsErr . Just $
+      "The requested point is too far in the past and has been pruned"
+  Core.AheadOfLastSync mResult ->
+    mkJsonRpcInvalidParamsErr . Just $
+      "The required point is ahead of the current index"
+        <> case mResult of
+          Nothing -> ""
+          Just r -> ". The latest result was: " <> show r
+  Core.IndexerQueryError msg ->
+    mkJsonRpcInternalErr . Just $
+      "The indexer query failed: " <> Text.unpack msg
```

Comment:
We may like a pretty instance for errors that we would use it, although it can be done in a separate ticket.

Reply:
Let's do it in a separate ticket